proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+
+Pros: High efficiency, minimal performance impact, no need to add additional instruction in transaction,
+helps to avoid denial of service attack on the smart contract.
+
+Cons: Modification of account structure which impacts a lot of code. Needs to load all accounts to calculate
+total fees to charge the payer, to avoid that have to implement multiple level of fees.
+
+## New Terminology
+
+Application Fees: Fees that are decided by the dapp developer that will be charged if a user wants
+to use the dapp. They will be applied even if transaction fails contrary to lamport transfers.
+
+## Detailed Design
+
+As a high-performance cluster, we want to incentivize players which feed the network
+with correctly formed transactions instead of spamming the network. The introduction
+of application fees would be an interesting way to penalize the bad actors and dapps
+can rebate these fees to the good actors. This means the dapp developer has to decide who to
+rebate and who to penalize special instructions. In particular, it needs to be able to
+penalize, even if the application is not CPI'd to. There were multiple possible approaches
+discussed to provide the application with access to transfer lamports outside of the regular cpi
+execution context. The following approach seems the best.
+
+*Passing the application fees in the instruction for each account and validating in the dapp*. A `PayApplicationFee`
+instruction is like irrevocable transfer instruction and will do the transfer even if the transaction fails. A new
+instruction `CheckApplicationFee` will check if the application fee for an account has been paid.
+
+    1. High degree of flexibility for programs (**++**)
+    2. No need to save the application fees on the ledger. (**++**)
+    3. Each transaction has to include additional instructions so it will break existing dapp interfaces (**-**)
+    4. Account structure does not need to be modified (**++**)
+    5. Does not prevent any denial of service attack on a smart contract (**-**)
+    6. Each dapp has to implement a cpi to check if a transaction has paid app fees
+
+If existing dapps like openbook want to implement application fees then all the dapps depending on openbook 
+also have to do additional development. The checks on the application fees will be responsibility of dapp developers.
+The plus point is that we do not have to touch the core solana structure to store the application fees.
+It is also more easier to calculate total fees.
+
+### A new application fee program
+
+We add a new native solana program called application fees program with program id.
+
+```
+App1icationFees1111111111111111111111111111
+```
+
+This native program will be used to check application fee for an account, to intialize rebates
+initaited by the account authority, and a special instruction by which fee payer accepts
+amount of lamports they are willing to pay as application fees per account.
+
+#### PayApplicationFees Instruction
```

Comment:
My main gripe with this is that write lock fees appear to be the discretionary decision of the DApp developer as opposed to something that's dynamically computed by the L1 as a function of demand for state. Correct me if I'm wrong, but it seems like the way the fees to get updated is for the DApp developer to essentially update the `CheckApplicationFees` instruction (or any of its inputs that determine the fee price)

While I think fees should certainly adjust based off of demand for state, I don't think the runtime has all of the information you need in order to make a system like this truly market driven. Namely, I think the runtime would need to be able to tell you how much compute a particular account has used up in a slot (in a variable, not just through `sol_log_compute_units`). It probably would also want to know about the current "price" to lock this account as well as the fees collected in a particular block.

If the mechanism for adjusting app fees could have a default implementation based off of block space demand I think this could be reasonable. However, without that, I think the solution seems centralizing and potentially inefficient at pricing block demand

Reply:
There are other proposals of dynamic fees which address your concerns. This proposal does not aims address the price a block as per demand issue.

This proposal address as a dapp developer to charge a fixed fee to interact with the dapp, or to whitelist or blacklist some transaction.

It also aims to reduce spamming of the dapp by encouraging creation of valid transactions by looking at the state of the cluster.

Reply:
@godmodegalactus I can completely respect that this proposal has one set of side effects but I think the argument here is that the other proposals aim to accomplish the same task. What @jarry-xiao and others are getting at (and correct me if I am wrong @jarry-xiao) is that dynamic fees solve the same problems. If constructed correctly and with consideration for supply and demand for blockspace and write lock contention, dapp developers should not need this punch through into the L1 fee economics.

Reply:
I beg to differ, dynamic fees will increase fees for everyone without any rebate mechanism there is no mechanism to punish malicious users. By this proposal, we do not punish well-intended users.
I think solana still wants to be a network where gas fees are low. With dynamic fees based on write lock the whole dapp will be punished if these fees are high and nothing will change if the fees are low.

Other than that it will be hard to know for a payer how much fees they are going to pay when they sign the transaction, with dynamic fees there may be a huge difference between when the payer signed for a transaction and when the payer pays. If fees changes all the time users may loose confidence over the network. There is a lot of stuff to figure out before we can implement correctly `CalculateFeeForMessage` and synchronize TPU and TVU so that they calculate the same fees.